xline-client/src/clients/lock.rs

```diff
@@ -16,26 +30,275 @@ pub struct LockClient {
     name: String,
     /// The client running the CURP protocol, communicate with all servers.
     curp_client: Arc<CurpClient<Command>>,
+    /// The lease client
+    lease_client: LeaseClient,
     /// The watch client
     watch_client: WatchClient,
     /// Auth token
     token: Option<String>,
 }
 
+// These methods primarily originate from xline lock server,
+// see also: `xline/src/server/lock_server.rs`
 impl LockClient {
-    /// New `LockClient`
+    /// Creates a new `LockClient`
     #[inline]
     pub fn new(
         name: String,
         curp_client: Arc<CurpClient<Command>>,
         channel: Channel,
         token: Option<String>,
+        id_gen: Arc<LeaseIdGenerator>,
     ) -> Self {
         Self {
-            name,
-            curp_client,
+            name: name.clone(),
+            curp_client: Arc::clone(&curp_client),
+            lease_client: LeaseClient::new(
+                name,
+                curp_client,
+                channel.clone(),
+                token.clone(),
+                id_gen,
+            ),
             watch_client: WatchClient::new(channel, token.clone()),
             token,
         }
     }
+
+    /// Acquires a distributed shared lock on a given named lock.
+    /// On success, it will return a unique key that exists so long as the
+    /// lock is held by the caller. This key can be used in conjunction with
+    /// transactions to safely ensure updates to Xline only occur while holding
+    /// lock ownership. The lock is held until Unlock is called on the key or the
+    /// lease associate with the owner expires.
+    ///
+    /// # Errors
+    ///
+    /// If `CurpClient` fails to send request
+    #[inline]
+    pub async fn lock(&self, request: LockRequest) -> Result<LockResponse> {
```

Comment:
I think we can use a `LockGuard` and implement its `Drop` trait to unlock automatically, instead of just returning a wrapper of some data. This design enforces [RAII](https://doc.rust-lang.org/rust-by-example/scope/raii.html) style in Rust.